Add lp-ocp-compat-cr-onboarding agent skill (#45)
CSPI-QE/ocp-ci-docs: .cursor/skills/lp-ocp-compat-cr-onboarding/SKILL.md

---
name: lp-ocp-compat-cr-onboarding
description: >-
  Onboard a Layered Product (LP) OCP Compatibility CI Operator Job into
  Component Readiness (CR) across openshift/release, openshift/sippy, and
  openshift-eng/ci-test-mapping. Use when executing LP OCP Compat CR onboarding,
  opening CR PRs, or when invoked as /lp-ocp-compat-cr-onboarding with parameters.
disable-model-invocation: true
argument-hint: >-
  lp-name=My-product lp-slug=my-product lp-repo=myorg/myrepo lp-branch=main

>   lp-ver=lpGA ocp-release=4.22 release-config=ci-operator/config/... test-variant=aws

Reviewer: The official term is CI Conf. file, not `release-config`.
```suggestion
  lp-ver=lpGA ocp-release=4.22 ci-config=ci-operator/config/... test-variant=aws
```

  gh-user=USER make-maintainer=requester|agent|user [jira-component=LP--My-product]
paths:
  - docs/OCP_CI_Tutorials/Reporting/**
  - .cursor/skills/lp-ocp-compat-cr-onboarding/**
---

# LP OCP Compat Component Readiness onboarding

> Onboard a Layered Product (LP) into Component Readiness (CR) via three PRs in **openshift/release**, **openshift/sippy**, and **openshift-eng/ci-test-mapping**.

Reviewer 2: **No motivation or demand signal**

The skill automates a multi-repo workflow but never states how often onboarding happens or how long the manual process takes. Consider adding a brief motivation so future readers understand the ROI.

```suggestion
Onboard a Layered Product (LP) into Component Readiness (CR) via three PRs in **openshift/release**, **openshift/sippy**, and **openshift-eng/ci-test-mapping**. This skill replaces a manual multi-repo workflow that is error-prone (case sensitivity, identifier derivation, cross-repo coordination).
```

Authoritative file-edit rules: [Reporting Guide, Component Readiness](../../../docs/OCP_CI_Tutorials/Reporting/Reporting_Guide.md#component-readiness).

Full mock PR content for the MPEXOperator proof of concept: [references/worked-example-mpexoperator.md](references/worked-example-mpexoperator.md).

## Before starting

> When `$ARGUMENTS` is present, parse each `key=value` token from the invocation (see `argument-hint` above). Required keys: `lp-name`, `lp-slug`, `lp-repo`, `lp-branch`, `lp-ver`, `ocp-release`, `release-config`, `test-variant`, `gh-user`, `make-maintainer`. Optional: `jira-component` (defaults to `LP--<lp-name>`). Treat `make-maintainer=user` as `requester`.

Reviewer 2: **10 required parameters impose high cognitive load**

Several parameters appear derivable: `lp-slug` from `lp-name`, `lp-branch` nearly always `main`, `lp-ver` nearly always `lpGA`. Reducing required inputs lowers the adoption barrier.

```suggestion
When `$ARGUMENTS` is present, parse each `key=value` token from the invocation (see `argument-hint` above). Required keys: `lp-name`, `lp-repo`, `ocp-release`, `release-config`, `test-variant`, `gh-user`, `make-maintainer`. Optional (with defaults): `lp-slug` (defaults to lowercased `lp-name`), `lp-branch` (defaults to `main`), `lp-ver` (defaults to `lpGA`), `jira-component` (defaults to `LP--<lp-name>`). Treat `make-maintainer=user` as `requester`.
```

> When `$ARGUMENTS` is present, parse each `key=value` token from the invocation (see `argument-hint` above). Required keys: `lp-name`, `lp-slug`, `lp-repo`, `lp-branch`, `lp-ver`, `ocp-release`, `release-config`, `test-variant`, `gh-user`, `make-maintainer`. Optional: `jira-component` (defaults to `LP--<lp-name>`). Treat `make-maintainer=user` as `requester`.

Reviewer: ```suggestion
When `$ARGUMENTS` is present, parse each `key=value` token from the invocation (see `argument-hint` in the YAML frontmatter). Required keys: `lp-name`, `lp-slug`, `lp-repo`, `lp-branch`, `lp-ver`, `ocp-release`, `release-config`, `test-variant`, `gh-user`, `make-maintainer`. Optional: `jira-component` (defaults to `LP--<lp-name>`). Treat `make-maintainer=user` as `requester`.
```

If any required input is missing, stop and list what is needed.

> Derive the **Derived identifiers** table from the [Reporting Guide, Onboarding Inputs](../../../docs/OCP_CI_Tutorials/Reporting/Reporting_Guide.md#onboarding-inputs) patterns and the validated inputs; do not invent values.

Reviewer 2: **Nit — terminology drift**
"Derived identifiers" here vs. "identifier table" in the Pull requests section and worked example.

```suggestion
Derive the **identifier table** from the [Reporting Guide, Onboarding Inputs](../../../docs/OCP_CI_Tutorials/Reporting/Reporting_Guide.md#onboarding-inputs) patterns and the validated inputs; do not invent values.
```

For the MPEXOperator proof of concept, when only product name, OCP release, fork owner, and `make-maintainer` are known, use the remaining parameter values from [Worked Example: MPEXOperator](references/worked-example-mpexoperator.md#invocation).

## At a glance

> - **Phase 0:** temp `$WORKDIR`; clone `release`, `sippy`, `ci-test-mapping`; remotes `upstream` (official) and `origin` (fork); sync `main`; feature branch per repo.

Reviewer: Use full `org/repo` notation please.

Following our Best Practices, please use the canonical form  (`${WORKDIR}`) when referring to shell variables.

AI skill is supposedly use natural language, `temp foo` doesn't sound like natural language.
```suggestion
- **Phase 0:** temp. dir. `${WORKDIR}`; clone `release`, `sippy`, `ci-test-mapping`; remotes `upstream` (official) and `origin` (fork); sync `main`; feature branch per repo.
```

> - **release first:** CR-compliant CI Operator Job Conf.; verify JUnit Test Suite (TS) prefix in Prow Job artifacts before other PRs.

Reviewer 2: **"before other PRs" is temporally ambiguous**

Does this mean before creating or before merging? Conflicts with the cross-linking requirement that all three PR URLs exist in every body.

```suggestion
- **release first:** CR-compliant CI Operator Job Conf.; verify JUnit Test Suite (TS) prefix in Prow Job artifacts before merging sippy and ci-test-mapping PRs.
```

Reviewer: Please use the full `org/repo` notation:
```suggestion
- **`openshift/release` PR:** CR-compliant CI Operator Job Conf.; verify JUnit Test Suite (TS) prefix in Prow Job artifacts (via Job Rehearsal) before creating `openshift-eng/sippy` and `openshift-eng/ci-test-mapping` PRs.
```

- **sippy:** `setLayeredProduct`, `config/views.yaml`, variant snapshot (Steps 1, 2, 6 usually skipped for standard LP OCP Compat).
- **ci-test-mapping:** component package and registry (Step 1 usually skipped).
- **Deliverables:** upstream PR URLs, local run log and shell trace (default `.cursor/skills/lp-ocp-compat-cr-onboarding/runs/`), cleanup `$WORKDIR`.

Every commit message must cite **`[ocp-ci-docs] lp-ocp-compat-cr-onboarding`**.

## Phase 0 workspace

> ```bash

Reviewer: Please follow our best practices when writing shell scripts. This issue has occurred several times now; we need to shift our paradigm to avoid unnecessary PR review feedback cycles.

WORKDIR=$(mktemp -d -t cr-onboarding-XXXXXX)
cd "$WORKDIR"

> [ ! -d release ] && git clone https://github.com/openshift/release.git

Reviewer 2: **Full clone of openshift/release (~1 GB)**

A full clone downloads the entire history, which in an agent environment might time out. The skill only edits a single file; a shallow clone is sufficient.

```suggestion
[ ! -d release ] && git clone --depth 1 https://github.com/openshift/release.git
```

Reviewer: Thank you, this is a VERY GOOD suggestion. Better compliment with single branch clone:
```suggestion
[ ! -d release ] && git clone --depth=1 --single-branch --no-tags https://github.com/openshift/release.git
```

[ ! -d sippy ] && git clone https://github.com/openshift/sippy.git
[ ! -d ci-test-mapping ] && git clone https://github.com/openshift-eng/ci-test-mapping.git
```

> After first clone in each repo: rename `origin` to `upstream`; add fork as `origin` (`https://github.com/<gh-user>/<repo>.git`). Sync `main` from `upstream`; create feature branch (example `mpexoperator`).

Reviewer 2: **Fork existence not verified before push**

`git remote add` succeeds regardless of whether the remote exists; the failure surfaces at `git push` with no recovery path.

```suggestion
After first clone in each repo: rename `origin` to `upstream`; verify fork exists with `gh repo view <gh-user>/<repo> --json name` (create with `gh repo fork` if missing); add fork as `origin` (`https://github.com/<gh-user>/<repo>.git`). Sync `main` from `upstream`; create feature branch (example `mpexoperator`).
```

Reviewer: It is probably best not to require gh.
Instead, we should clone the fork directly to verify that it exists. This means the fork URL should be passed as an argument to the skill rather than assuming a default format like `https://github.com/<gh-user>/<repo>.git` for everyone. From there, we can configure the `upstream` remote.

```bash
typeset workDir="$(mktemp -d -t cr-onboarding-XXXXXX)"
typeset -A rmtURLs=(
    openshift/release=<openshift/release_ForkURL>
    openshift/sippy=<openshift-eng/sippy_ForkURL>
    openshift-eng/ci-test-mapping=<openshift-eng/ci-test-mapping_ForkURL>
)

pushd "${workDir}"
for upRmt in "${!rmtURLs[@]}"; do
    [ -d "${upRmt#*/}" ] || git clone --depth=1 --single-branch --no-tags "${rmtURLs[${upRmt}]}"
    pushd "${upRmt#*/}"
    git remote get-url upstream &>/dev/null || git remote add upstream "https://github.com/${upRmt}.git"
    git remote set-url --push upstream noPush
    git switch "$(git rev-parse --abbrev-ref upstream/HEAD | sed 's|^upstream/||')"
    git fetch --depth=1 --no-tags -pP origin HEAD
    git fetch --shallow-exclude=HEAD --update-shallow --no-tags -pP upstream HEAD
    git reset --hard upstream/HEAD
    git push --force-with-lease origin HEAD
    git switch -C cr-onboarding--mpex-operator
    popd
done
popd
```

If Git cannot run, stop and report what is needed; do not apply unverified edits.

> ## Implementation order

Reviewer 2: **No strategy for upstream repo convention drift**

The skill hardcodes naming conventions for 3 repos the team does not own. When upstream changes conventions, the skill silently generates incorrect edits.

```suggestion
## Implementation order

Before generating edits, verify that expected anchor patterns exist in the cloned repos (e.g., `setLayeredProduct` in sippy, `pkg/components/` in ci-test-mapping, CI Operator config directory layout in release). Halt with a clear message if the structure does not match expectations.
```

```mermaid
flowchart LR
  release[openshift/release]
  sippy[openshift/sippy]

>   ctm[ci-test-mapping]

Reviewer: ```suggestion
  ctm[openshift-eng/ci-test-mapping]
```

  release --> sippy
  release --> ctm
```

### openshift/release

See [CI Operator Job Configuration](../../../docs/OCP_CI_Tutorials/Reporting/Reporting_Guide.md#ci-operator-job-configuration).

> - CI Operator Job full name must contain `-<lpVer>-lp-ocp-compat-cr--<lpName>-`.

Reviewer 2: **Template variable `<lpName>` maps to `lp-slug`, not `lp-name`**

The worked example reveals `<lpName>` = `mpexoperator` (lowercase slug). Using the parameter name removes the ambiguity.

```suggestion
- CI Operator Job full name must contain `-<lpVer>-lp-ocp-compat-cr--<lp-slug>-`.
```

Reviewer: The master onboarding document uses `lpName`. To avoid confusion, we should probably use `lpName` as the parameter key instead of `lp-slug`.

> - `.tests[].steps.env`: `MAP_TESTS: "true"`, `DR__RP__CR_COMP_NAME: lp-ocp-compat--<LP-name>`.

Reviewer 2: **Template variable `<LP-name>` maps to `lp-name`**

Same issue as above: `<LP-name>` = `MPEXOperator` (display case `lp-name`). Using the parameter name removes ambiguity.

```suggestion
- `.tests[].steps.env`: `MAP_TESTS: "true"`, `DR__RP__CR_COMP_NAME: lp-ocp-compat--<lp-name>`.
```

- `.tests[].cron` at least twice daily.
- ExitTrap / JUnit post-processing when the Test Step uses `mpiit-data-router-reporter` or `firewatch-ipi-aws-cr`.

### openshift/sippy

> See [Sippy](../../../docs/OCP_CI_Tutorials/Reporting/Reporting_Guide.md#sippy). Standard LP OCP Compat: add Step 3 `setLayeredProduct`, Step 4 `config/views.yaml`, Step 5 snapshot; skip Steps 1, 2, 6.

Reviewer 2: **Sippy lowercase transformation not documented**

Sippy's `setLayeredProduct` lowercases the job name before matching against `layeredProductPatterns`. The worked example shows this implicitly (`lpGA` → `lpga`), but the rule is never stated. A novel LP would produce a mixed-case substring that silently fails to match.

```suggestion
See [Sippy](../../../docs/OCP_CI_Tutorials/Reporting/Reporting_Guide.md#sippy). Standard LP OCP Compat: add Step 3 `setLayeredProduct`, Step 4 `config/views.yaml`, Step 5 snapshot; skip Steps 1, 2, 6. The `layeredProductPatterns` substring must be all-lowercase (`lpVer` and `lpSlug` lowercased) because `setLayeredProduct` matches against `strings.ToLower(jobName)`.
```

Reviewer: We shouldn't repeat what is already stated in the reference documentation. The AI Agent is expected to read the reference and understand that the `layeredProductPatterns` substrings must be lowercase.

### openshift-eng/ci-test-mapping

See [CI Test Mapping](../../../docs/OCP_CI_Tutorials/Reporting/Reporting_Guide.md#ci-test-mapping). Standard LP OCP Compat: add `pkg/components/<lpComp>/` and registry entry; skip `includeSuitePatterns` when `lp-ocp-compat--%` already applies.

## Pull requests

- Push feature branch to fork `origin`.
- Open PR on upstream with `--head <gh-user>:<branch>`.

> - Cross-link all three upstream PR URLs in every PR body; include identifier table and maintainer `make` lines not yet run.

Reviewer 2: **Cross-linking requires back-updates not described**

PRs are created sequentially, so the first can't have URLs for PRs that don't exist yet.

```suggestion
- Cross-link all three upstream PR URLs in every PR body; include identifier table and maintainer `make` lines not yet run. After all PRs are open, back-update each body with the full set of cross-links via `gh pr edit`.
```

Example:

```bash
gh pr create --repo openshift/sippy --head <gh-user>:<branch> --base main \
  --title "Onboard MPEXOperator for LP OCP Compat Component Readiness (Sippy)"
```

## Maintainer `make`

> | Repo            | Command                                                              | When `make-maintainer` is `requester` or `user` |

Reviewer 2: **`make-maintainer=agent` mode is advertised but undefined**

The `argument-hint` offers `requester|agent|user` but the table only documents `requester`/`user` behavior. An agent encountering `make-maintainer=agent` would not know whether to run the commands, commit the results, or both, and the environmental prerequisites (Go toolchain, repo dependencies) are unstated.

Replace lines 103-107 with:

```markdown
| Repo            | Command                                                              | When `make-maintainer` is `requester` or `user` | When `make-maintainer` is `agent`  |
|-----------------|----------------------------------------------------------------------|-------------------------------------------------|------------------------------------|
| sippy           | `make update-variants`; `./sippy variants snapshot --config ./config/openshift.yaml` | Document in PR body only                        | Run, commit regenerated artifacts  |
| ci-test-mapping | `make mapping`                                                       | Document in PR body only                        | Run, commit regenerated artifacts  |
| release         | `make update` / `make jobs`                                          | Document in PR body only                        | Run, commit regenerated artifacts  |
```

Also document the prerequisite environment for `agent` mode (Go toolchain, `make`, repo-specific dependencies).

|-----------------|----------------------------------------------------------------------|-------------------------------------------------|
| sippy           | `make update-variants`; `./sippy variants snapshot --config ./config/openshift.yaml` | Document in PR body only                        |
| ci-test-mapping | `make mapping`                                                       | Document in PR body only                        |

> | release         | `make update` / `make jobs`                                          | Document in PR body only                        |

Reviewer: Allign please.

- **Deliverables:** upstream PR URLs, local run log and shell trace (default `.cursor/skills/lp-ocp-compat-cr-onboarding/runs/`), cleanup `$WORKDIR`.

Every commit message must cite **`[ocp-ci-docs] lp-ocp-compat-cr-onboarding`**.

## Phase 0 workspace

```bash
#!/bin/bash
set -euxo pipefail; shopt -s inherit_errexit

typeset workDir
workDir="$(mktemp -d -t cr-onboarding-XXXXXX)"
typeset -A rmtURLs=(
    ["openshift/release"]="https://github.com/<gh-user>/release.git"
    ["openshift/sippy"]="https://github.com/<gh-user>/sippy.git"
    ["openshift-eng/ci-test-mapping"]="https://github.com/<gh-user>/ci-test-mapping.git"
)

pushd "${workDir}"
typeset upRmt mainBranch
for upRmt in "${!rmtURLs[@]}"; do
    [ -d "${upRmt#*/}" ] || git clone --depth=1 --single-branch --no-tags "${rmtURLs[${upRmt}]}"
    pushd "${upRmt#*/}"
    git remote get-url upstream &>/dev/null || git remote add upstream "https://github.com/${upRmt}.git"
    git remote set-url --push upstream noPush
    mainBranch="$(git rev-parse --abbrev-ref upstream/HEAD | sed 's|^upstream/||')"
    git switch "${mainBranch}"
    git fetch --depth=1 --no-tags -pP origin HEAD
    git fetch --depth=1 --no-tags -pP upstream HEAD

>     git reset --hard upstream/HEAD

Reviewer 2: **`upstream/HEAD` is not resolvable after manual `remote add`.** Three compounding issues here:

1. `git rev-parse --abbrev-ref upstream/HEAD` fails because git only creates `refs/remotes/<remote>/HEAD` during `git clone` or via `git remote set-head` ([git-remote docs](https://git-scm.com/docs/git-remote/2.37.0), [RFC PATCH discussion](https://public-inbox.org/git/20241203215713.135068-1-bence@ferdinandy.com/T/): "`clone` sets `refs/remotes/[remote]/HEAD` while going `init -> remote add -> fetch` does not").
2. `git fetch upstream HEAD` stores the commit in `FETCH_HEAD` but does not create `refs/remotes/upstream/main`, so `git remote set-head upstream -a` would also fail (the docs say "this will only work if `refs/remotes/origin/next` already exists; if not it must be fetched first").
3. `git reset --hard upstream/HEAD` on line 70 hits the same missing ref.

Using `git ls-remote --symref` to query the default branch (no fetch needed), then fetching the specific branch name to create proper tracking refs, avoids all three issues.

```suggestion
    mainBranch="$(git ls-remote --symref upstream HEAD | sed -nE 's|^ref: refs/heads/(\S+)\s+HEAD$|\1|p')"
    : "${mainBranch:=main}"
    git fetch --depth=1 --no-tags -pP origin "${mainBranch}"
    git fetch --depth=1 --no-tags -pP upstream "${mainBranch}"
    git switch "${mainBranch}"
    git reset --hard "upstream/${mainBranch}"
```

Author: Thank you, this thread made me re-think about this, I have simplified the entire script.

---
name: lp-ocp-compat-cr-onboarding
description: >-
  Onboard a Layered Product (LP) OCP Compatibility CI Operator Job into
  Component Readiness (CR) across openshift/release, openshift/sippy, and
  openshift-eng/ci-test-mapping. Use when executing LP OCP Compat CR onboarding,
  opening CR PRs, or when invoked as /lp-ocp-compat-cr-onboarding with parameters.
disable-model-invocation: true
argument-hint: >-
  lp-name=My-product lp-slug=my-product lp-repo=myorg/myrepo lp-branch=main

>   lp-ver=lpGA ocp-release=4.22 ci-config=ci-operator/config/... test-variant=aws

Reviewer 2: **Parameter naming inconsistency.** This uses `ci-config` but line 28 and the worked example use `release-config`. Worth aligning on one name to avoid confusion.

```suggestion
  lp-ver=lpGA ocp-release=4.22 release-config=ci-operator/config/... test-variant=aws
```

  cron="0 6,18 * * *" gh-user=USER make-maintainer=requester|agent|user [jira-component=LP--My-product]
paths:
  - docs/OCP_CI_Tutorials/Reporting/**

>   - .cursor/skills/lp-ocp-compat-cr-onboarding/**

Reviewer 2: **`disable-model-invocation: true` makes `paths:` inert.** With model invocation disabled, the skill only fires on explicit `/lp-ocp-compat-cr-onboarding`. The `paths:` block has no auto-trigger effect. Is this intentional (e.g., for discoverability)? If not, could be removed to reduce confusion.

```suggestion
disable-model-invocation: true
argument-hint: >-
  lp-name=My-product lp-slug=my-product lp-repo=myorg/myrepo lp-branch=main
  lp-ver=lpGA ocp-release=4.22 release-config=ci-operator/config/... test-variant=aws
  cron="0 6,18 * * *" gh-user=USER make-maintainer=requester|agent|user [jira-component=LP--My-product]
```

- **Deliverables:** upstream PR URLs, local run log and shell trace (default `.cursor/skills/lp-ocp-compat-cr-onboarding/runs/`), cleanup `$WORKDIR`.

Every commit message must cite **`[ocp-ci-docs] lp-ocp-compat-cr-onboarding`**.

## Phase 0 workspace

```bash
#!/bin/bash
set -euxo pipefail; shopt -s inherit_errexit

typeset workDir
workDir="$(mktemp -d -t cr-onboarding-XXXXXX)"
typeset -A rmtURLs=(
    ["openshift/release"]="https://github.com/<gh-user>/release.git"
    ["openshift/sippy"]="https://github.com/<gh-user>/sippy.git"
    ["openshift-eng/ci-test-mapping"]="https://github.com/<gh-user>/ci-test-mapping.git"
)

pushd "${workDir}"
typeset upRmt mainBranch
for upRmt in "${!rmtURLs[@]}"; do
    [ -d "${upRmt#*/}" ] || git clone --depth=1 --single-branch --no-tags "${rmtURLs[${upRmt}]}"
    pushd "${upRmt#*/}"
    git remote get-url upstream &>/dev/null || git remote add upstream "https://github.com/${upRmt}.git"
    git remote set-url --push upstream noPush
    mainBranch="$(git rev-parse --abbrev-ref upstream/HEAD | sed 's|^upstream/||')"
    git switch "${mainBranch}"
    git fetch --depth=1 --no-tags -pP origin HEAD

>     git fetch --shallow-exclude=HEAD --update-shallow --no-tags -pP upstream HEAD

Reviewer 2: **`--shallow-exclude` expects a named ref (branch/tag), not `HEAD`.** This may error on most git versions. If the intent is to fetch only the tip from upstream, `--depth=1` would work consistently.

```suggestion
    git fetch --depth=1 --no-tags -pP upstream HEAD
```
